src/data/list/cycle.lean

```diff
@@ -716,4 +722,107 @@ underlying element. This representation also supports cycles that can contain du
 instance [has_repr α] : has_repr (cycle α) :=
 ⟨λ s, "c[" ++ string.intercalate ", " ((s.map repr).lists.sort (≤)).head ++ "]"⟩
 
+/-- `chain R s` means that `R` holds between adjacent elements of `s`.
+
+`chain R ([a, b, c] : cycle α) ↔ R a b ∧ R b c ∧ R c a` -/
+def chain (r : α → α → Prop) (c : cycle α) : Prop :=
+quotient.lift_on' c (λ l, match l with
+  | [] := true
+  | (a :: m) := chain r a (m ++ [a]) end) $
+λ a b hab, propext $ begin
+  cases a with a l;
+  cases b with b m,
+  { refl },
+  { have := is_rotated_nil_iff'.1 hab,
+    contradiction },
+  { have := is_rotated_nil_iff.1 hab,
+    contradiction },
+  { unfold chain._match_1,
+    cases hab with n hn,
+    induction n with d hd generalizing a b l m,
+    { simp only [rotate_zero] at hn,
+      rw [hn.1, hn.2] },
+    { cases l with c s,
+      { simp only [rotate_singleton] at hn,
+        rw [hn.1, hn.2] },
+      { rw [nat.succ_eq_one_add, ←rotate_rotate, rotate_cons_succ, rotate_zero, cons_append] at hn,
+        rw [←hd c _ _ _ hn],
+        simp [and.comm] } } }
+end
+
+@[simp] lemma chain.nil (r : α → α → Prop) : cycle.chain r (@nil α) :=
+by trivial
+
+@[simp] lemma chain_coe_cons (r : α → α → Prop) (a : α) (l : list α) :
+  chain r (a :: l) ↔ list.chain r a (l ++ [a]) :=
+iff.rfl
+
+@[simp] lemma chain_singleton (r : α → α → Prop) (a : α) : chain r [a] ↔ r a a :=
+by rw [chain_coe_cons, nil_append, chain_singleton]
+
+lemma chain_ne_nil (r : α → α → Prop) {l : list α} :
+  Π hl : l ≠ [], chain r l ↔ list.chain r (last l hl) l :=
+begin
+  apply l.reverse_rec_on,
+  exact λ hm, hm.irrefl.elim,
+  intros m a H _,
+  rw [←coe_cons_eq_coe_append, chain_coe_cons, last_append]
+end
+
+lemma chain_map {β : Type*} {r : α → α → Prop} (f : β → α) {s : cycle β} :
+  chain r (s.map f) ↔ chain (λ a b, r (f a) (f b)) s :=
+quotient.induction_on' s $ λ l, begin
+  cases l with a l,
+  refl,
+  convert list.chain_map f,
+  rw map_append f l [a],
+  refl
+end
+
+variables {r : α → α → Prop} {s : cycle α}
+
+theorem chain_of_pairwise : (∀ (a ∈ s) (b ∈ s), r a b) → chain r s :=
```

Comment:
I would guess you can move `hs` before the colon here now that you're using the `induction` tactic